### PR TITLE
Workaround for the failure of ConvHipImplicitGemmV4R4GenWrWXdlops

### PR DESCRIPTION
**Pull request by zjing14**
Resolves issue #406

### CODE REVIEW

**Comment by zjing14** [comment]
Tested with configs from the issue. No performance drop after workaround.

File: src/solver/implicitgemm_util.hpp [hunk 1]
```diff
@@ -23,6 +23,8 @@ MIOPEN_DECLARE_ENV_VAR(MIOPEN_DEBUG_CONV_IMPLICIT_GEMM_BLOCK_SYNC_LDS_WITHOUT_SY
 // LLVM xdlops instrinsic will do unnecessey VGRP <--> AGPR movement, and result in
 // register spill, for bfloat16 datatype, when doing wave-wise GEMM larger than 64x64
 #define WORKAROUND_SWDEV_240356 1
+// workaround failure of ConvHipImplicitGemmV4R4GenWrWXdlops with vector load
+#define WORKAROUND_ISSUE_2532 1
```
Comment (atamazov) [inline src/solver/implicitgemm_util.hpp:27]: [Recommendation] Used only once in a .cpp? Difine right there.

**Review by atamazov — APPROVED** [review]
LGTM

**Comment by atamazov** [comment]
Topmost comment & priority fixed.

**Comment by ekuznetsov139** [comment]
"No performance drop after workaround"

That sounds wrong to me. In my tests, I've seen performance drops as high as 50%.

**Comment by atamazov** [comment]
@ekuznetsov139 Perhaps re-tuning is required. @zjing14 What do you think?

**Comment by zjing14** [comment]
@ekuznetsov139 Could you post the config with regression here? @atamazov No, I did not retune.

**Comment by ekuznetsov139** [comment]
MIOpenDriver convfp16 -n 256 -c 512 -H 28 -W 28 -k 128 -y 1 -x 1 -p 0 -q 0 -u 1 -v 1 -l 1 -j 1 -m conv -g 1 -F 4 -t 1

Without the fix:
stats: name, n, c, ho, wo, x, y, k, flopCnt, bytesRead, bytesWritten, GFLOPs, GB/s, timeMs
stats: bwdw-conv1x1u1, 256, 512, 28, 28, 1, 1, 128,  26306674688, 0, 0, 29434, 0, 0.893762
Backward Convolution Weights Failed: 0.241877 > 0.082

With the fix:
stats: name, n, c, ho, wo, x, y, k, flopCnt, bytesRead, bytesWritten, GFLOPs, GB/s, timeMs
stats: bwdw-conv1x1u1, 256, 512, 28, 28, 1, 1, 128,  26306674688, 0, 0, 14505, 0, 1.813586
Backward Convolution Weights Verifies on CPU and GPU (4.11355e-05)

**Comment by zjing14** [comment]
@ekuznetsov139 Thanks. Sorry, I did not compare the performance of failed configs, since I think it does not fair. Yes, that is huge performance degradation.

**Comment by daniellowell** [comment]
@zjing14  So, this is a WIP again?

**Comment by zjing14** [comment]
> @zjing14  So, this is a WIP again?

No, the PR is ready.

**Comment by daniellowell** [comment]
@zjing14 What is the plan to reduce the impact of the performance regression? How widespread is this regression?

**Comment by zjing14** [comment]
@daniellowell This solver will be deprecated after the new wrw solver merged in. So, the impact is temporary.

**Comment by atamazov** [comment]
@zjing14 
> @ekuznetsov139 Thanks. Sorry, I did not compare the performance of failed configs, since I think it does not fair. Yes, that is huge performance degradation.

You were correct. The notion of performance is not applicable to the kernels that produce wrong outputs.

However, currently find-db contains too optimistic information about `ConvHipImplicitGemmV4R4GenWrWXdlops`, especially where find-db records say that it wins. That might lead to performance drops, **_unless_** `ConvHipImplicitGemmV4R4GenWrWXdlops` remains the winner even after workaround applied (or at least on par with the real winner). Does it?

**Pull request merged**